### ISSUE TITLE
Timer inconsistency issues

### ISSUE THREAD

**Issue by gprasant** [builtwithluv/ZenFocus#208]
Bug report #206 seems to report a similar issue and has a good template for bug reports. So copying the template from there :-)

**Environment**
macOS Sierra 10.12.6
ZenFocus Version 1.4.0

**Description**
[Timer inconsistencies Video](https://www.dropbox.com/s/dsg52o4k81dfhlj/zenfocus_timer.mov?dl=0)

**Observations**
I managed to reliably reproduce this issue through the following steps:
1. Launch ZenFocus and start a session. It is counting down at a normal pace now. 
2. Click on the ZenFocus menu button and select Resume. 
3. Observe the timer. It is counting down at a much faster pace (perhaps double.)

**Expected Result**
The timer is accurate to the system clock. It is either running (at a constant pace) or not running depending on the state of the UI controls.

**UPDATES** 
[2017-10-13] - Updated the Observations to include the reliable steps to reproduce the issue.

**Comment by chengsieuly** [comment]
Thanks! Yes, any extra information would be great We'll look into this issue =)

**Comment by chengsieuly** [comment]
I tried to replicate the issue but it's not happening. I also dug through the code and can't find out where that could happen. Please let me know once you know what user actions would lead to the behavior!

**Comment by ryanbaer** [comment]
Let's make this the representative issue of this bug (as opposed to mine in #206).

Correct me if I'm wrong, but it seems to me that the count down timer in this app functions through the use of setInterval and dispatching actions to decrement the timer in the UI.

While it's perfectly fine to update the timer in the UI by dispatching actions, I suspect the inaccuracies are stemming from the assumption that you can count on Javascript timers to be accurate. My understanding is that they're notorious for being inaccurate.

Keep in mind that Electron is just embedded Chromium, so underlying this is simply a webpage. And while I haven't confirmed this, I think that Chromium may be detecting that none of the windows are active, and thus de-prioritizing the subprocess containing the timer - hence, your setInterval is not running every 1000ms, but much less frequently, when the window is minimized.

Take a look at the following article which details the inaccuracies of these functions:
https://www.sitepoint.com/creating-accurate-timers-in-javascript/

The first example indicates why you couldn't rely on setInterval to make a stopwatch, for example.

So if you can't count on setInterval and setTimeout to run with the accuracy you'd like, what can you count on? The standard approach I've seen, and what is demonstrated in the article, is to instead measure the current timestamp relative to the previous timestamp to determine how much time has elapsed. 

From the article:
```
var start = new Date().getTime(),
    elapsed = '0.0';

window.setInterval(function()
{
    var time = new Date().getTime() - start;

    elapsed = Math.floor(time / 100) / 10;
    if(Math.round(elapsed) == elapsed) { elapsed += '.0'; }

    document.title = elapsed;

}, 100);
```

To implement this in your app, with `setSeconds` as an example, instead of dispatching:
```
dispatch(setSeconds(seconds - 1));
```

You would dispatch:
```
dispatch(setSeconds(seconds - deltaSeconds));
```
where `deltaSeconds` is the elapsed number of seconds since the last time the function ran.

Implemented in your app, this would result in an accurate timer in which the UI maybe skips a second or two from time to time (0:25 to 0:23, for example). Or perhaps it wouldn't even be noticeable to that degree.

**Comment by ryanbaer** [comment]
Additionally, I do remember seeing @gprasant's issue of the timer not stopping once or twice. 

I haven't looked into how you implemented pause logic in the UI, so maybe you already addressed this. But could it be that, when the user hits pause, actions have already been dispatched to decrement the counter, so it continues to work through those actions, decrementing the counter even after paused?

Another explanation could be that, for whatever reason, the function containing `clearInterval` isn't getting called in some circumstances.

Also, perhaps it would be helpful to do the following:

- When calling `clearInterval(ticker)`, set `ticker` to null or a value indicating that the interval has been cleared.
- In `resume`, confirm that `ticker` is indeed null before calling `setInterval` again.

The reason I suggest this is because @gprasant's video makes it appear as if multiple `setInterval` functions are running simultaneously, whereas your app operates off of the assumption that the setInterval called on tick should effectively be a singleton, with only one existing at any given time.

**Comment by chengsieuly** [comment]
Your assumptions are very correct. For whatever reason, it may be that on pause, the current set interval is not being cleared. Doing what you suggested - assuring ticker is null on clear and assuring there is no other ticker on resume could resolve the issue when it happens. It boggles me still why it would happen in the first place. I can’t seem to pinpoint it nor replicate it myself either.

**Comment by gprasant** [comment]
@chengsieuly  I managed to find out a sequence of user actions to reliably reproduce the issue. Updated the description.

**Comment by chengsieuly** [comment]
@gprasant sweet! Thanks. I will look into this

**Comment by ehguo** [comment]
While it seems like there are a number of issues relating to the timing implementation, this particular one seems like it's just making sure you can't `resume` while the counter is already in countdown mode yea?

**Comment by gprasant** [comment]
Yeah. I'd venture a solution to make Pause and Resume menus mutually exclusive and display only the relevant one.

**Comment by gprasant** [comment]
Awesome. The issue appears to be fixed in 1.4.1. Thanks @chengsieuly

**Comment by chengsieuly** [comment]
@gprasant Sweet! Please let us know if there are any other issues that you find or any ideas you think would give users a better experience =)
